### ISSUE TITLE
AI同士の対戦をlive観戦できるSpectator modeを既存GUIへ追加する

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong-play#25]
Parent architecture decision: https://github.com/lisbun/lisjong-project/issues/7
Related durable record consumer: https://github.com/lisbun/lisjong-arena/issues/155

# Status

**PLANNED — live spectator consumer / no dependency on Arena #155.**

`lisjong-play` には現在、first-party `lisjong-engine` を使う Human EAST vs selected Policy x3 のTkinter GUIがあり、viewer-relative table、public scores、rivers、melds、riichi state、dora indicators、round metadata、Human hand、局結果、半荘結果まで表示できる。

本Issueではこの既存GUIを再利用し、**Human操作を持たない AI x4 対戦を人間がlive観戦できるSpectator mode** を追加する。

目的は新しいViewer applicationを別実装することではなく、既存の盤面・結果表示をhuman-play modeとspectator modeで共有できるようにすることである。

---

# Goal

conceptual flow:

```text
AI Policy x4
    ↓
first-party lisjong-engine execution
    ↓
live spectator presentation source
    ↓
existing lisjong-play GUI presentation
    ↓
human spectator
```

SpectatorはActionを選択しない。

最低限、AI4席で1半荘を完走しながら、対局進行をGUI上で観戦できるようにする。

---

# Primary user experience

起動時にHuman Playとは明確に区別されたSpectator modeを選択できること。

初期scopeで最低限表示する:

- round / honba / riichi sticks等の局情報
- 4席のscore / wind / dealer state
- rivers
- melds
- riichi state
- dora indicators
- current turn / recent actionを理解するために必要な進行表示
- 各局の終了結果
- 半荘終了時の最終score / rank

可能なら4席のconcealed handをspectator-only表示する。

ただしconcealed-hand表示のために:

- engine内部mutable stateを直接読む
- private implementation fieldへGUIをbindする
- 4席のhidden informationをPolicy-visible stateへ混ぜる
- GUI都合のomniscient objectをPolicy入力へ渡す

ことは禁止する。

current `lisjong-engine` に適切なengine-owned spectator / objective projectionが存在しない場合は、private-state readで回避せず、その不足をbounded prerequisiteとして切る。

---

# Execution / Policy boundary

AI席はreal `lisjong.Policy` を使用し、既存のfirst-party Policy bridgeを可能な限りthin reuseする。

初期modeは既存CLI/GUIのPolicy selection semanticsに合わせ、4席すべて同じselected Policyでもよい。

per-seat arbitrary Policy composition / league UIは本Issueの要件にしない。

Policy instanceはseatごとに独立させる。

Spectator UIはPolicy判断ロジック、合法手判定、麻雀ruleを所有しない。

---

# Presentation architecture

Human PlayとSpectatorで盤面rendererを二重実装しない。

現行GUIのview / rendering codeを起点に、必要な範囲だけlive presentation sourceとの差を分離する。

conceptual target:

```text
Human Play source ──┐
                    ├─ shared board/result presentation
Spectator source ───┘
```

ただし本Issueで `GenericViewerFramework` やproject-wide `ViewerState` contractを先行設計しない。

Spectatorを成立させるために実際に重複するpresentation logicだけを抽出する。

将来の牌譜Viewerはこのshared presentationを再利用する予定だが、本Issue内でReplay sourceは実装しない。

---

# Playback controls

live観戦向けに最低限次を提供する。

```text
Pause
Resume
Step (one presentation/action boundary)
Speed
```

exact UI placement / keyboard shortcutはcurrent GUI conventionsへ合わせる。

## Pause semantics

pause中もgame stateを壊さず、次のpresentation boundaryへの進行だけを止める。

UI threadでengine stateを直接mutateしない。

## Step semantics

pause中に1単位だけ進められること。

単位は実装前にcurrent engine/session boundaryを確認して明示する。複数の内部eventを曖昧に1 actionと呼ばず、GUI側でruleを推測しない。

## Speed

presentation delay / pacingだけを変更し、Policy result、seed、engine semanticsを変えない。

speed変更で同じseedの対局結果が変わらないことを可能な範囲で固定する。

---

# Threading / lifecycle

現在のTkinter GUIはTk main threadとengine workerを分離している。この原則を維持する。

- Tk widgetはmain threadだけで操作
- engine / Policy executionをTk callback内で長時間blockingしない
- window closeでworker / pending pauseを解放できる
- pause / step / speed controlがdeadlockを作らない
- exceptionをcatch-allで成功扱いしない

Human selector向けの待機機構をSpectatorへ無理に偽装せず、AI-only executionに自然なpresentation gateを作る。

---

# Hidden information boundary

Spectatorはhuman playerではないため、観戦用途ではconcealed informationの表示を許容できる。

ただし情報flowを明示的に一方向にする。

```text
engine-owned authoritative / spectator-safe projection
        ↓
lisjong-play presentation

Policy-visible seat observation
        ↓
Policy
```

Spectator側hidden informationをPolicy側へ戻さない。

AI decision comparison / training label / HandBelief ground truthをViewer内で作らない。

---

# Result presentation

既存の局結果 / 半荘結果presentationを可能な限り再利用する。

Human confirmationがないため、Spectatorでは:

- 局結果を表示
- pause / configurable presentation delay等のboundedな進行方法
- 次局へ自動または明示的なspectator controlで進行

のいずれかをcurrent GUI structureに合わせて設計してよい。

Human Playのconfirmation semanticsを壊さない。

---

# Preflight

実装開始時に最低限確認する。

- current Issue / latest comments
- `lisbun/lisjong-project#7`
- `lisjong-play` current main / `AGENTS.md`
- `README.md`
- `docs/architecture.md`
- current Tkinter GUI entry point / board renderer / view model
- current Human selector / worker-thread bridge
- round-result / hanchan-result presentation
- current AI Policy bridge and Policy factory resolution
- `lisjong-engine` のavailable observation / progress / spectator-like projection
- AI x4 session compositionをexisting public contractだけで成立させられるか

4 concealed handsを表示するためのpublic/engine-owned projectionが無い場合、private state accessを追加せずre-scopeまたはbounded engine prerequisiteを提案する。

---

# Tests

## Session / execution

- AI4席でHuman selectorを要求せずhalf-game completionできる
- Policyはseatごとにfresh instance
- same seed / same Policy compositionでspectator pacing on/offがsemantic resultを変えない
- spectator modeがlegal action / scoring / progressionを再実装しない

## GUI / controls

- spectator modeでhuman action panelを表示しない、またはinteractive human action controlとして機能しない
- pause / resumeが進行を安全に止め・再開する
- paused stepがpredefined 1 presentation boundaryだけ進める
- speed changeがpresentation pacingだけを変更する
- window closeでworkerを残さない
- error時にfinished resultを捏造しない

## Presentation

- existing score / river / meld / riichi / dora / round metadata rendererを再利用する
- 局結果 / 半荘結果を表示する
- hidden-hand表示を行う場合、spectator-only sourceから来ておりPolicy inputへ混入しない

real hanchanを大量に増やさず、fake/focused testsと既存fixed-seed integrationを優先する。

---

# Documentation

最低限更新候補:

- `README.md`
- `docs/architecture.md`

明記する:

```text
Human Play GUI
    human seat + AI seats

Spectator GUI
    AI x4 + no human action selection

Replay Viewer
    separate successor consumer
```

---

# Non-goals

- Arena #155 durable record persistence
- replay / 牌譜Viewer
- save / resume
- database
- web UI
- network spectator protocol
- RiichiLab live観戦
- Mortal match viewer
- per-seat arbitrary Policy league editor
- AI strength evaluation
- Policy reasoning / logits / Q-value visualization
- training data generation
- viewer-side shanten / ukeire / scoring / legality calculation
- canonical ViewerState / generic frontend framework
- project-wide GameRecord
- tile artwork / animation / sound polishをcompletion requirementにすること

---

# Acceptance criteria

- [ ] current GUI / engine / Policy bridgeをpreflightした
- [ ] Human Playとは別にSpectator modeを起動できる
- [ ] AI x4で1 half-gameをHuman inputなしに完走できる
- [ ] existing board/result presentationを再利用している
- [ ] 4席のscore / rivers / melds / riichi / dora / round progressをlive表示できる
- [ ] current turn / recent actionを観戦者が追える
- [ ] human action selection UIをSpectatorで要求しない
- [ ] Pause / Resume / bounded Step / Speedが成立する
- [ ] pacing変更がgame semanticsへ影響しない
- [ ] 局結果を表示する
- [ ] 半荘結果 / final rankingを表示する
- [ ] concealed handsを表示する場合はengine-owned safe projectionを利用し、private mutable stateを読んでいない
- [ ] spectator hidden informationがPolicy-visible inputへ逆流しない
- [ ] Tk main thread / worker boundaryを維持する
- [ ] window close / error pathでdeadlockやsuccessful-result偽装をしない
- [ ] Viewer側でrule / scoring / legality / Policy reasoningを再実装していない
- [ ] Replay / durable persistence / generic Viewer frameworkへscopeを広げていない
- [ ] affected tests / docsを更新している

# Expected successor

牌譜Viewerは別Issueで、Arena #155のstrict-loaded durable recordをsourceとして本Issueで成立したpresentationを再利用する。

**Comment by lisbun** [comment]
## Successor link — Replay / 牌譜Viewer

Live Spectatorの後継consumerとして `lisbun/lisjong-play#26` を起票した。

役割分担は次で固定する。

```text
#25
live AI x4 spectator source
    ↓
existing/shared presentation

#26
Arena #155 durable record
    ↓
replay presentation source
    ↓
existing/shared presentation
```

#25ではReplay source / durable persistenceを実装しない。#26では#25で実際に成立したpresentation reuseを優先するが、両Issueを理由にgeneric Viewer frameworkを先行設計しない。

**Comment by lisbun** [comment]
## 2026-09-12 status refresh

This Issue remains valid, but the presentation baseline has advanced since the body was written.

```text
Human Play GUI       implemented
Replay Viewer #26    implemented / merged via PR #33
shared GuiBoardRenderer established
Arena durable record schema v2 available
Spectator #25        still PLANNED
```

Therefore Replay Viewer is no longer a future successor consumer. Spectator mode should become the **third presentation source** reusing the already-shared board renderer:

```text
Human Play source ──┐
Replay source ───────┼─> GuiBoardRenderer / shared presentation
Spectator source ────┘
```

This does not make Arena #155/#207 persistence a dependency for live spectator execution. Spectator still needs a live engine-owned presentation source, and any four-seat concealed-hand display must use an explicit spectator/objective projection rather than private mutable engine state.

At implementation preflight, inspect current Replay Viewer / `GuiBoardRenderer` first and extract no new generic viewer framework unless concrete spectator differences require it.
